pimd, pim6d: Fix RPF check

patrasar opened the pull request:
1. Zebra should send ifindex  for nexthop type NEXTHOP_TYPE_IPV6.

2. pim6d should handle nexthop type NEXTHOP_TYPE_IPV6.

3. Fix RPF check,
For a  IPV4/IPV6 route, nexthop can be primary or secondary address.
Whereas PIM hello packet source IP is always primary address.
When there is a difference in the RPF address and PIM neighbour
address for the same interface. RPF check fails.
Fix: 
Check RPF address is either neighbour primary address  or one of the secondary address list.
   
Closes #11957
Closes #11526
Closes #12103

ton31337 commented:
ci:rerun CI maintenance

davischw commented:
ci:rerun restarted after dequeued due to ci maintenance

mobash-rasool commented on pimd/pim_nht.c:
```diff
@@ -576,6 +612,15 @@ static int pim_ecmp_nexthop_search(struct pim_instance *pim,
 			pim_addr nhaddr = nh_node->gate.ipv6;
 #endif
 			nbrs[i] = pim_neighbor_find(ifps[i], nhaddr);
+#if PIM_IPV == 6
```
Since whenever pim_neighbor_find api fails you are trying to call pim_neighbor_find_by_secondary function everywhere, I think it will be better to move this section of code inside pim_neighbor_find api so you don't need to change code wherever the api is called. This will avoid duplication of code as well.

patrasar replied:
Addressed.

mobash-rasool commented on pimd/pim_rpf.c:
```diff
@@ -128,6 +128,16 @@ bool pim_nexthop_lookup(struct pim_instance *pim, struct pim_nexthop *nexthop,
 			   && !pim_if_connected_to_source(ifp, addr)) {
 			nbr = pim_neighbor_find(ifp,
 						nexthop_tab[i].nexthop_addr);
+#if PIM_IPV == 6
```
All these code can be removed if we put inside pim_neighbor_find api.

mobash-rasool commented:
Changes looks fine now but I feel the commits can be more properly organised. Better to club 1st and 2nd commit together and 3rd and 4th together.

patrasar commented:
> Changes looks fine now but I feel the commits can be more properly organised. Better to club 1st and 2nd commit together and 3rd and 4th together.

Thanks Mobash.
As discussed offline, all are different independent issues. Thats the reason I kept different commits.
Let's sync up, if anything needed.

eqvinox commented:
Uh… neighbors should only send PIM packets with their primary link local address as source — are you sure there isn't something broken on the TX side?

patrasar commented:
> Uh… neighbors should only send PIM packets with their primary link local address as source — are you sure there isn't something broken on the TX side?

Hi David,
As per rfc7761 section 4.9,
PIM messages are either unicast (e.g., Registers and Register-Stop)
or multicast with TTL 1 to the 'ALL-PIM-ROUTERS' group (e.g.,
Join/Prune, Asserts). The source address used for unicast messages
is a domain-wide reachable address; the source address used for
multicast messages is the link-local address of the interface on
which the message is being sent.
As per rfc5059 section 4,
The IP source address used for Bootstrap messages
(regardless of whether they are being originated
or forwarded) is the link-local address of the
interface on which the message is being sent.

I raised https://github.com/FRRouting/frr/pull/12167,  which will take care of transmitting side by not allowing the Multicast message with global address. But there is no where mentioned in the RFC that, we should process or drop multicast packet with global address. So closed the PR as of now.

eqvinox commented:
> I raised #12167, which will take care of transmitting side by not allowing the Multicast message with global address. But there is no where mentioned in the RFC that, we should process or drop multicast packet with global address. So closed the PR as of now.

That PR was actually better; there are 2 reasons to reject packets in receive:

- if we accept GUA sources, it's a security problem because PIM on-link packets can now be sent from some random place in the internet. If another vendor gets it wrong and sends on-link PIM packets with a global source, they need to fix that, we can't weaken security for their bugs. (Yes, Register and Candidate-RP messages need to be accepted from GUA, but that does not mean the check can be dropped for everything.)
- if we accept LL addresses other than the primary LL address, it can break DR election. I'd rather not take that risk.

⇒ really only primary LL address should be accepted. (same for IPv4, except primary address is not LL.)

Only Register and Candidate-RP (not currently implemented) messages use GUA sources, so only those 2 should be excluded from the check.

Can you update #12167 appropriately?

patrasar commented:
> > I raised #12167, which will take care of transmitting side by not allowing the Multicast message with global address. But there is no where mentioned in the RFC that, we should process or drop multicast packet with global address. So closed the PR as of now.
> 
> That PR was actually better; there are 2 reasons to reject packets in receive:
> 
> * if we accept GUA sources, it's a security problem because PIM on-link packets can now be sent from some random place in the internet. If another vendor gets it wrong and sends on-link PIM packets with a global source, they need to fix that, we can't weaken security for their bugs. (Yes, Register and Candidate-RP messages need to be accepted from GUA, but that does not mean the check can be dropped for everything.)
> * if we accept LL addresses other than the primary LL address, it can break DR election. I'd rather not take that risk.
> 
> ⇒ really only primary LL address should be accepted. (same for IPv4, except primary address is not LL.)
> 
> Only Register and Candidate-RP (not currently implemented) messages use GUA sources, so only those 2 should be excluded from the check.
> 
> Can you update #12167 appropriately?

Sure @eqvinox.
I will open the PR #12167. We can discuss more in that.

Here, I have fixed the issue where "PIMD should be able to do RPF check properly, when Zebra sends nexthop which can be either link-local or global address".

So requesting you to review this one and merge :)

patrasar commented:
I have addressed the above comment in https://github.com/FRRouting/frr/pull/12167, Here we need to check if the Join/Prune, assert packet is received from primary link-local address of the neighbour. This is done by calling the API "pim_neighbor_find".

I can't move pim_neighbor_find_by_secondary inside pim_neighbor_find() API, so keeping this separate.

patrasar commented:
ci:rerun

riw777 reviewed (approved):
looks fine to me ... might be good to have another set of eyes here before merging, though

ton31337 reviewed (approved):
LGTM, just not sure if we need a topotest here?

mobash-rasool reviewed (approved):
LGTM

ton31337 commented:
@Mergifyio backport dev/8.5
